tapchannel/aux_sweeper.go

```diff
@@ -360,6 +384,190 @@ func commitRevokeSweepDesc(keyRing *lnwallet.CommitmentKeyRing,
+// called after a force close to ensure that we can properly spend outputs
+// created by the commitment transaction at a later step.
+func (a *AuxSweeper) importCommitTx(req lnwallet.ResolutionReq,
+	commitState *cmsg.Commitment, fundingInfo *cmsg.OpenChannel) error {
+
+	// To start, we'll re-create vPackets for all of the outputs of the
+	// commitment transaction.
+	//
+	// We'll use the fundingInfo proofs to create a vIn for each of them.
+	fundingInputProofs := make(map[asset.ID]*proof.Proof)
+	for _, fundingInput := range fundingInfo.FundedAssets.Val.Outputs {
+		inputProof := &fundingInput.Proof.Val
+		fundingInputProofs[inputProof.Asset.ID()] = inputProof
+	}
+
+	// With the funding proof for each asset ID known, we can now make the
+	// vPackets for each of the outputs.
+	vPktsByAssetID := make(map[asset.ID]*tappsbt.VPacket)
+	for _, localAsset := range commitState.LocalAssets.Val.Outputs {
+		err := assetOutputToVPacket(
+			fundingInputProofs, vPktsByAssetID, localAsset,
+			&a.cfg.ChainParams, req.CommitTx,
+			a.cfg.DefaultCourierAddr,
+		)
+		if err != nil {
+			return err
+		}
+	}
+	for _, remoteAsset := range commitState.RemoteAssets.Val.Outputs {
+		err := assetOutputToVPacket(
+			fundingInputProofs, vPktsByAssetID, remoteAsset,
+			&a.cfg.ChainParams, req.CommitTx,
+			a.cfg.DefaultCourierAddr,
+		)
+		if err != nil {
+			return err
+		}
+	}
+	for _, outgoingHTLCs := range commitState.OutgoingHtlcAssets.Val.HtlcOutputs {
+		for _, outgoingHTLC := range outgoingHTLCs.Outputs {
+			err := assetOutputToVPacket(
+				fundingInputProofs, vPktsByAssetID, outgoingHTLC,
+				&a.cfg.ChainParams, req.CommitTx,
+				a.cfg.DefaultCourierAddr,
+			)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for _, incomingHTLCs := range commitState.IncomingHtlcAssets.Val.HtlcOutputs {
+		for _, incomingHTLC := range incomingHTLCs.Outputs {
+			err := assetOutputToVPacket(
+				fundingInputProofs, vPktsByAssetID, incomingHTLC,
+				&a.cfg.ChainParams, req.CommitTx,
+				a.cfg.DefaultCourierAddr,
+			)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// With all the vPackets created, we'll create output commitments from
+	// them, as we'll need them to ship the transaction off to the porter.
+	vPkts := maps.Values(vPktsByAssetID)
+	ctx := context.Background()
+	for idx := range vPkts {
+		err := tapsend.PrepareOutputAssets(ctx, vPkts[idx])
+		if err != nil {
+			return fmt.Errorf("unable to prepare output "+
+				"assets: %w", err)
+		}
+	}
+	outCommitments, err := tapsend.CreateOutputCommitments(vPkts)
+	if err != nil {
+		return fmt.Errorf("unable to create output "+
+			"commitments: %w", err)
+	}
+
+	// Before we ship things off, we'll also insert all the script keys
+	// (which have a tap leaf) associated with them into the db. This way
+	// we'll be able to recognize the output as belonging to the wallet.
+	//
+	// TODO(roasbeef): remote's output known, but only when it's a breach
```

Comment:
Can we detect which ones are ours and which ones aren't? Otherwise we'll have unspent assets in our DB forever (because currently we don't mark assets as spent if their anchor UTXO is spent, only if the proof for spending the asset is imported into the same node again).

Reply:
Yeah I ended up resolving this for that very reason. Now we'll only import the outputs that we _know_ are ours.